prov/rxm/src/rxm_conn.c

```diff
@@ -686,27 +676,35 @@ void rxm_cmap_free(struct rxm_cmap *cmap)
 	struct dlist_entry *entry;
 	size_t i;
 
+	rxm_cmap_cm_thread_close(cmap);
+
+	/* cleanup function would be used in manual progress mode */
+	if (cmap->ep->domain->data_progress != FI_PROGRESS_AUTO)
+		rxm_conn_cleanup(cmap->ep);
+
 	cmap->acquire(&cmap->lock);
+
 	FI_DBG(cmap->av->prov, FI_LOG_EP_CTRL, "Closing cmap\n");
 	for (i = 0; i < cmap->num_allocated; i++) {
-		if (cmap->handles_av[i])
-			rxm_cmap_del_handle(cmap->handles_av[i]);
+		if (cmap->handles_av[i]) {
+			rxm_cmap_clear_key(cmap->handles_av[i]);
+			rxm_conn_free(cmap->handles_av[i]);
+			cmap->handles_av[i] = 0;
+		}
 	}
 	while(!dlist_empty(&cmap->peer_list)) {
 		entry = cmap->peer_list.next;
 		peer = container_of(entry, struct rxm_cmap_peer, entry);
-		rxm_cmap_del_handle(peer->handle);
+		dlist_remove(&peer->entry);
+		rxm_cmap_clear_key(peer->handle);
+		rxm_conn_free(peer->handle);
+		free(peer);
 	}
-	rxm_cmap_cm_thread_close(cmap);
-	cmap->release(&cmap->lock);
-
-	/* cleanup function would be used in manual progress mode */
-	if (cmap->ep->domain->data_progress != FI_PROGRESS_AUTO)
-		rxm_conn_cleanup(cmap->ep);
 
 	free(cmap->handles_av);
 	free(cmap->attr.name);
 	ofi_idx_reset(&cmap->handles_idx);
+	cmap->release(&cmap->lock);
```

Comment:
We free cmap and the lock immediately after releasing the lock.  This usually indicates that the lock isn't needed, or we're going to have a use-after-free crash anyway.  Looking at the changes and quickly checking the existing code, I think it's the former.  The cleanup looks good.